connect-inject/health_check_resource_test.go

```diff
@@ -0,0 +1,606 @@
+
+	// Start the clientset with a Pod that is failed
+	pod := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      testPodName,
+			Namespace: "default",
+			Labels:    map[string]string{labelInject: "true"},
+			Annotations: map[string]string{
+				annotationStatus:  "injected",
+				annotationService: testServiceNameAnnotation,
+			},
+		},
+		Spec: testPodSpec,
+		Status: corev1.PodStatus{
+			HostIP: "127.0.0.1",
+			Phase:  corev1.PodRunning,
+			Conditions: []corev1.PodCondition{{
+				Type:   corev1.PodReady,
+				Status: corev1.ConditionFalse,
+			}},
+		},
+	}
+	k8sclientset := fake.NewSimpleClientset(pod)
+	randomPorts := freeport.MustTake(6)
+	schema := "http://"
+	serverAddress := fmt.Sprintf("%s%s:%d", schema, "127.0.0.1", randomPorts[1])
+
+	// setup consul client connection
+	clientConfig := &api.Config{Address: serverAddress}
+	require.NoError(err)
+	client, err := api.NewClient(clientConfig)
+	require.NoError(err)
+	consulUrl, err := url.Parse(serverAddress)
+	require.NoError(err)
+
+	healthResource := HealthCheckResource{
+		Log:                 hclog.Default().Named("healthCheckResource"),
+		KubernetesClientset: k8sclientset,
+		ConsulUrl:           consulUrl,
+		ReconcilePeriod:     100 * time.Millisecond,
+	}
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	defer cancelFunc()
+
+	// start the reconciler
+	go func() {
+		healthResource.Run(ctx.Done())
+	}()
+	// let reconcile run at least once
+	time.Sleep(time.Millisecond * 300)
+
+	testServerReady := make(chan bool)
+	var srv *testutil.TestServer
+	go func() {
+		srv, err = testutil.NewTestServerConfigT(t, func(c *testutil.TestServerConfig) {
+			c.Ports = &testutil.TestPortConfig{
+				DNS:     randomPorts[0],
+				HTTP:    randomPorts[1],
+				HTTPS:   randomPorts[2],
+				SerfLan: randomPorts[3],
+				SerfWan: randomPorts[4],
+				Server:  randomPorts[5],
+			}
+		})
+		require.NoError(err)
+		close(testServerReady)
+	}()
+	// Wait for server to come up
+	select {
+	case <-testServerReady:
+		defer srv.Stop()
+	}
+	// validate that there is no health check created by reconciler
+	check := getConsulAgentChecks(t, client)
+	require.Nil(check)
+	// Add the service - only now will a health check have a service to register against
+	srv.AddService(t, testServiceNameReg, api.HealthPassing, nil)
+
+	// retry so we can cover time period when reconciler is already running vs
+	// when it will run next based on the loop
+	timer := &retry.Timer{Timeout: 5 * time.Second, Wait: 1 * time.Second}
+	var actual *api.AgentCheck
+	retry.RunWith(timer, t, func(r *retry.R) {
+		actual = getConsulAgentChecks(t, client)
+		// the assertion is not on actual != nil, but below
+		// against an expected check.
+		if actual == nil {
+			r.Error("check = nil")
+		}
+	})
+
+	expectedCheck := &api.AgentCheck{
+		CheckID: testHealthCheckID,
+		Status:  api.HealthCritical,
+		Notes:   testFailureMessage,
+		Output:  testFailureMessage,
+	}
+	// Validate the checks are set
```

Comment:
You can populate the expectedCheck with Name and Type and perform a `require.Equal(expectedCheck, actual)` here to simplify things